### DISCUSSION TITLE
Syntactic sugar for reflection?

### DISCUSSION THREAD

**Discussion by Flithor** [dotnet/csharplang#4159] [Language Ideas] [answered]
Just an idea:
`$(Class)` equals to `typeof(class)` get the class type
`$(obj)` equals to `obj?.GetType()` get the object type
`$(Class.Property)` equals to `typeof(class).GetProperty("Property")` get the ProepertyInfo
and more?
Sometimes I think if we can got the PropertyInfo like get Type by `typeof(Class)` will be great.

**Comment by JanTamis** [comment]
I don't think that that is a good idea because people will be more tempted to use reflection, which is slow. Also i don't think that it is clear what the syntax does to someone who doesn't know the syntax.

**Reply by Phantonia** [reply]
I don't think a language should make a feature complicated to use just because it is slow. Every developer should decide for themselves (or their team obviously), which features to use and which features not to use. Where I agree though is that the syntax is not really clear. A keyword instead of the dollar character would make it clearer, imo.

**Reply by orthoxerox** [reply]
@Phantonia The language should make better choices easier to use than worse choices. Reflection should generally be the option of last resort if you can't use the type system to achieve the same goal. That's why it's clunkier to use than the rest of the language and hasn't received any major API updates: use just enough of it to solve the problem, wrap what you've written in a better API, return to regular C#.

Take a look at source generators. They are another form of metaprogramming and aren't blessed with a super-streamlined API either. Again, this nudges you towards the idea that these tools are not expected to be used by every C# programmer there is, but by "tool makers" that create finished libraries for wider consumption. This, of course, isn't a universal truth at all, but a stance taken by the designers of C#, compare this with something like Common Lisp, where it is expected that every team will use its metaprogramming facilities to create a language specific to the problem domain, so macros are easy to write and have good syntax support.

**Comment by canton7** [comment, marked as answer]
1. What's so painful about `typeof(C)` or `c.GetType()` that it needs to be reduced by a few characters?
2. How do you handle the overloads of `GetProperty` which take more than just a string?
3. How do you handle the other `GetXXX` methods, e.g. for methods/events/etc? Why does `$(Class.Foo)` get you a `PropertyInfo`, and not a `MethodInfo` or `EventInfo`, etc?
4. C# has `dynamic` which may help with some of these cases. E.g. `typeof(C).GetProperty("Foo").Invoke(c)` is similar to `((dynamic)c).Foo`

**Reply by Flithor** [reply]
100+ `nameof(Class.Property)` as hard code for get PropertyInfo, they will use for mapping something.
We owned typeof to get a type object in code. I think if we owned something new like `propof`, `methodof` operator for get them in code conveniently is a great idea.
Yes, I want more meta things.

**Reply by CyrusNajmabadi** [reply]
> 100+ nameof(Class.Property) as hard code for get PropertyInfo

This is already legal today and means something else m this would be a breaking change
